### PR TITLE
Handle HTML comments

### PR DESCRIPTION
**Pull request by feliperenan**
This PR not only improves the Formatter to handle  HTML comments but also improve the code a bit. Here are the main changes:

* Handle HTML comments. 
* Build the HTML using a `buffer` list and `IO.io_data_to_binary` instead of appending with `<>`. 
* Some internal changes such as renaming modules. Now we have `heex_formatter/formatter.ex` and `heex_formatter/tokenizer`. These modules are called by the `heex_formatter.ex` and that's all.

### CODE REVIEW

File: lib/heex_formatter/phases/tokenizer.ex [hunk 1]
```diff
@@ -27,46 +31,160 @@ defmodule HeexFormatter.Phases.Tokenizer do
...
+  defp do_tokenize({type, _, _, opt, expr}, {tokens, {:comment, _, _} = cont})
+       when type in @eex_expr do
+    expr = expr |> List.to_string() |> String.trim()
+
+    eex_tag =
+      if List.to_string(opt) == "=" do
+        "<%= #{expr} %>"
+      else
+        "<% #{expr} %>"
+      end
```
Comment (josevalim) [inline lib/heex_formatter/phases/tokenizer.ex]:
```suggestion
      "<%#{opt} #{expr} %>"
```

File: lib/heex_formatter/phases/tokenizer.ex [hunk 2]
```diff
@@ -27,46 +31,160 @@ defmodule HeexFormatter.Phases.Tokenizer do
...
+
+    # Get the last token which is the one the we want to append this eex_tag with.
```
Comment (paulo-silva) [inline lib/heex_formatter/phases/tokenizer.ex]:
```suggestion
    # Get the last token which is the one we want to append this eex_tag with.
```

File: lib/heex_formatter/phases/tokenizer.ex [hunk 3]
```diff
@@ -27,46 +31,160 @@ defmodule HeexFormatter.Phases.Tokenizer do
...
 
-  defp line_break?({:text, text, _meta}) do
+  @doc """
+  Finalize the Tokenization started by `Tokenizer.tokenize/1`.
+
+  This function is going to do some clean up and normalization to help the
+  Formatter module. Here are the final steps it does:
+
+  * Concatenate text from tags: script, tag, style, code, pre  and HTML comments
+    so that we don't we just append as it is in the Formatter.
```
Comment (paulo-silva) [inline lib/heex_formatter/phases/tokenizer.ex]:
```suggestion
  * Concatenate text from tags: script, tag, style, code, pre, and HTML comments
    we only append as it is in the Formatter.
```

File: lib/heex_formatter/phases/tokenizer.ex [hunk 4]
```diff
@@ -27,46 +31,160 @@ defmodule HeexFormatter.Phases.Tokenizer do
...
+
+  * Ignore line breaks that comes from HTMLTokenizer. That is because the Formatter
+    already know to handle line breanks.
+  """
+  def finalize(tokens) do
```
Comment (feliperenan) [inline lib/heex_formatter/phases/tokenizer.ex]: Ok, we might don't need it at all since we could handle this while tokenizing. This PR is supposed to help with that:

https://github.com/phoenixframework/phoenix_live_view/pull/1847

File: lib/heex_formatter/phases/format.ex [hunk 5]
```diff
@@ -85,9 +83,23 @@ defmodule HeexFormatter.Phases.Format do
...
+    %{state | html: state.html <> text}
+  end
+
+  defp token_to_string({:text, text, %{context: :comment_start}}, state) do
+    %{state | html: state.html <> String.trim_trailing(text), mode: :comment}
+  end
+
+  defp token_to_string({:text, text, %{context: :comment_end}}, state) do
+    %{state | html: state.html <> String.trim_trailing(text), mode: :normal}
   end
```
Comment (feliperenan) [inline lib/heex_formatter/phases/format.ex]: @josevalim The current code introduced in https://github.com/phoenixframework/phoenix_live_view/pull/1847 is enough to achieve what want here. It turns out that only adding `comment_start` and `comment_end` is enough since the script, among other tags we want to ignore, we handle by the tag itself because we know when it ends.

So now, we would have only two interactions here:
1. To build our tokens combining both `EExTonenizer` and `HTMLTokenizer`.
2. For formatting the code based on the tokens.

We might be able to format in just one go, but I guess it would make the code a bit hard to follow 🤔 

Please, let me know if that make sense 🙏🏻

Comment (josevalim) [reply lib/heex_formatter/phases/format.ex]: It is super natural to have multiple passes. so i would keep the structure from today. first group the tokens and then format.

File: lib/heex_formatter/formatter.ex [hunk 6]
```diff
@@ -85,9 +87,35 @@ defmodule HeexFormatter.Phases.Format do
         end
       end
 
+    mode = if tag in ~w(script tag style code pre), do: String.to_atom(tag), else: :normal
```
Comment (paulo-silva) [inline lib/heex_formatter/formatter.ex]: I would say to use to_existing_atom here, but we probably haven't use them previously
```suggestion
    mode = if tag in ~w(script tag style code pre), do: String.to_existing_atom(tag), else: :normal
```

Comment (paulo-silva) [reply lib/heex_formatter/formatter.ex]: actually we're defining them below, so probably to_existing_atom will work here?

Comment (feliperenan) [reply lib/heex_formatter/formatter.ex]: Not sure if we need it because we are pattern matching only these tags 🤔

File: lib/heex_formatter/formatter.ex [hunk 7]
```diff
@@ -17,42 +29,30 @@ defmodule HeexFormatter.Phases.Format do
...
-        new_state = token_to_string(token, state)
-
-        # Set the previous token so we can check it to know how the current tag should
-        # be formatted.
-        %{new_state | previous_token: token}
-      end)
+    tokens
+    |> Enum.reduce(initial_state, fn token, state ->
+      new_state = token_to_string(token, state)
+      %{new_state | previous_token: token}
```
Comment (feliperenan) [inline lib/heex_formatter/formatter.ex]: This is bothering me a bit because I guess I could store all tokens and pop the first to use as previous_token. But this also works and right now my current logic doesn't depend on a list of tokens but only the previous one. So I guess it is fine? 🤔

File: lib/heex_formatter/formatter.ex [hunk 8]
```diff
@@ -17,42 +29,30 @@ defmodule HeexFormatter.Phases.Format do
...
+    end)
+    |> buffer_to_string()
+  end
 
-    # That is because we are adding "\n" every time a tag is open. Then we need to extract
-    # "\n" from the first line and put this in the end of the line.
-    "\n" <> html = result.html
-    html <> "\n"
+  defp buffer_to_string(state) do
```
Comment (feliperenan) [inline lib/heex_formatter/formatter.ex:52]: Change the logic to not use `<>` but append items with `[left, right]` and apply `IO.io_data_to_binary` in the final result.

TIL:  We don't want to use `<>` within a loop to append binary 😅

File: lib/heex_formatter/formatter.ex [hunk 9]
```diff
@@ -343,17 +371,36 @@ defmodule HeexFormatter.Phases.Format do
...
   #   <p class="some-class">  \nShould break line
-  defp tag_contains_line_break?(html, tag) do
-    last_opened_tag = ~r/<\s*#{tag}[^>]*.(?!.*<\s*#{tag}[^>]*>(.*?).+).*/s
-
-    last_opened_tag
-    |> Regex.run(html)
-    |> case do
-      nil ->
-        false
+  defp tag_open_with_line_break?(buffer, tag) do
```
Comment (feliperenan) [inline lib/heex_formatter/formatter.ex:386]: Now we don't need regex anymore since we have a list of binaries now. So all we need to do here is to go to the tag open of the given tag and check if it contains a line break. I guess it will be faster than before 🤔

File: mix.exs [hunk 10]
```diff
@@ -21,7 +21,7 @@ defmodule HeexFormatter.MixProject do
   # Run "mix help deps" to learn about dependencies.
   defp deps do
     [
-      {:phoenix_live_view, "~> 0.17.5"},
+      {:phoenix_live_view, path: "~/source/phoenix_live_view"},
```
Comment (feliperenan) [inline mix.exs]: I will change it to LV master once my PR there gets merged.

File: lib/heex_formatter/formatter.ex [hunk 11]
```diff
@@ -86,8 +86,28 @@ defmodule HeexFormatter.Phases.Format do
...
+      else
+        text
+      end
+
+    %{state | buffer: [text | state.buffer], mode: mode}
+  end
 
-    %{state | html: state.html <> "\n" <> tag_opened, indentation: indentation}
+  defp token_to_string({tag, text, _meta}, %{mode: mode} = state)
+       when tag in ~w(text eex_tag eex_tag_render)a and mode in ~w(script style code pre comment)a do
```
Comment (paulo-silva) [inline lib/heex_formatter/formatter.ex]:
```suggestion
       when tag in ~w(text eex_tag eex_tag_render)a and mode in @special_modes do
```

File: lib/heex_formatter/formatter.ex [hunk 12]
```diff
@@ -343,17 +380,41 @@ defmodule HeexFormatter.Phases.Format do
...
+
+  defp may_add_line_break(:tag_open, {:text, text, _meta}) do
+    html_comment? = String.contains?(text, "<!--") and String.contains?(text, "-->")
+    if html_comment?, do: "", else: "\n"
+  end
+
+  defp may_add_line_break(:tag_open, _token), do: "\n"
+
+  # Returns `script`, `style`, `code` or `pre` when the given tag is one of these
+  # tag. Otherwise. it returns normal.
```
Comment (paulo-silva) [inline lib/heex_formatter/formatter.ex]:
```suggestion
  # tag. Otherwise, it returns `:normal`.
```

**Review by paulo-silva — APPROVED** [review]
Awesome 👏 - I only added minor comments 👍
